File: electron/main/crypto.ts
import CryptoJS from "crypto-js";
import crypto from "crypto";

export function generateSalt(): string {
  return crypto.randomBytes(32).toString("hex");
}

export function deriveKey(password: string, salt: string): string {
  return CryptoJS.PBKDF2(password, salt, {
    keySize: 256 / 32,
    iterations: 10000,
  }).toString();
}

export function encrypt(data: string, key: string): string {
  return CryptoJS.AES.encrypt(data, key).toString();
}

export function decrypt(encrypted: string, key: string): string {
  const decrypted = CryptoJS.AES.decrypt(encrypted, key);
  return decrypted.toString(CryptoJS.enc.Utf8);
}

export function hashPasswordForStorage(password: string): string {
  return CryptoJS.SHA256(password).toString();
}

export function hashNoteId(noteId: string): string {
  return CryptoJS.SHA256(noteId).toString().slice(0, 12);
}

File: cli-go/crypto.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"

	"golang.org/x/crypto/pbkdf2"
)

func deriveKey(password, salt string) []byte {
	return pbkdf2.Key([]byte(password), []byte(salt), 10000, 32, sha256.New)
}

func encrypt(plaintext string, keyHex []byte) string {
	key := keyHex[:32]

	block, err := aes.NewCipher(key)
	if err != nil {
		return ""
	}

	ciphertext := make([]byte, aes.BlockSize+len(plaintext))
	iv := ciphertext[:aes.BlockSize]
	if _, err := rand.Read(iv); err != nil {
		return ""
	}

	stream := cipher.NewCFBEncrypter(block, iv)
	stream.XORKeyStream(ciphertext[aes.BlockSize:], []byte(plaintext))

	return base64.StdEncoding.EncodeToString(ciphertext)
}

func decrypt(ciphertext []byte, keyHex []byte) (string, error) {
	key := keyHex[:32]

	data, err := base64.StdEncoding.DecodeString(string(ciphertext))
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	if len(data) < aes.BlockSize {
		return "", errors.New("ciphertext too short")
	}

	iv := data[:aes.BlockSize]
	data = data[aes.BlockSize:]

	stream := cipher.NewCFBDecrypter(block, iv)
	stream.XORKeyStream(data, data)

	return string(data), nil
}

func hashNoteID(noteID string) string {
	hash := sha256.Sum256([]byte(noteID))
	return hex.EncodeToString(hash[:])[:12]
}
